group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,399 @@
+            // If rack awareness is enabled, only add partitions if the consumers rack matches the partitions rack.
+            List<TopicIdPartition> validCurrentAssignment = getValidCurrentAssignment(memberId, assignmentMemberSpec.assignedPartitions());
+
+            int currentAssignmentSize = validCurrentAssignment.size();
+            int remaining = minQuota - currentAssignmentSize;
+
+            if (currentAssignmentSize > 0) {
+                int retainedPartitionsCount = min(currentAssignmentSize, minQuota);
+                for (int i = 0; i < retainedPartitionsCount; i++) {
+                    newAssignment.get(memberId)
+                        .targetPartitions()
+                        .computeIfAbsent(validCurrentAssignment.get(i).topicId(), k -> new HashSet<>())
+                        .add(validCurrentAssignment.get(i).partition());
+                    allAssignedStickyPartitions.add(validCurrentAssignment.get(i));
+                }
+
+                // The extra partition is located at the index "retainedPartitionsCount" from the current step.
+                if (remaining < 0 && remainingMembersToGetExtraPartition > 0) {
+                    newAssignment.get(memberId)
+                        .targetPartitions()
+                        .computeIfAbsent(validCurrentAssignment.get(retainedPartitionsCount).topicId(), k -> new HashSet<>())
+                        .add(validCurrentAssignment.get(retainedPartitionsCount).partition());
+                    allAssignedStickyPartitions.add(validCurrentAssignment.get(retainedPartitionsCount));
+                    remainingMembersToGetExtraPartition--;
+                }
+            }
+            if (remaining >= 0) {
+                potentiallyUnfilledMembers.put(memberId, remaining);
+            }
+        });
+
+        return allAssignedStickyPartitions;
+    }
+
+    /**
+     * Filters the current assignment of partitions for a given member.
+     *
+     * If a partition is assigned to a member not subscribed to its topic or
+     * if the rack-aware strategy is to be used but there is a mismatch,
+     * the partition is excluded from the valid assignment and stored for future consideration.
+     *
+     * @param memberId              The Id of the member whose assignment is being validated.
+     * @param assignedPartitions    The partitions currently assigned to the member.
+     *
+     * @return List of valid partitions after applying the filters.
+     */
+    private List<TopicIdPartition> getValidCurrentAssignment(String memberId, Map<Uuid, Set<Integer>> assignedPartitions) {
+        List<TopicIdPartition> validCurrentAssignmentList = new ArrayList<>();
+
+        assignedPartitions.forEach((topicId, currentAssignment) -> {
+            List<Integer> currentAssignmentList = new ArrayList<>(currentAssignment);
+            if (subscriptionList.contains(topicId)) {
```

Comment:
not necessarily, since there could be partitions belonging to old subscriptions or topics that don't exist anymore and we don't need them anymore aka they're not valid anymore which is fine. Mentioned in the javadoc

Reply:
seems like you've changed this to a set. is the comment above still valid?

Reply:
yes

Reply:
i'm not sure i follow. i've asked to change subscriptionList to a set, which you have done. what am i missing?

also, can we just use subscriptionSet instead of subscriptionList? it doesn't seem like we need a list to do any of the other operations.

Reply:
done